Change state correctly after migration finishes; make use of the Failed state

gjcolombo opened the pull request:
Ensure that after a live migration in or out, the state driver transitions the VM to some sensible state. #343 fixed some of this by making migration targets go to the Failed state after failing to migrate in; this PR aims to finish the job for all migration outcomes.

Turn the Failed state into a terminal state that destroys the underlying VM so that the control plane doesn't have to coordinate requests to stop such a VM. Make the VM controller's `Drop` impl preserve the Failed state so that it's distinct from Destroyed (sled agent/Nexus treat Destroyed as Stopped, but they shouldn't do this for Failed).

Tweak the migration code in the state driver to be crisper about who's responsible for publishing migration state updates. Previously, the state driver published some of these; now it assumes the migration tasks always do.

Fixes #209. Fixes #323. Fixes #326.

gjcolombo commented:
Strictly speaking, this PR implements a slightly different definition of "failed" than that described in #209. This PR uses it more or less to mean "we couldn't even start the VM," but #209 talks about having more of a "the guest has become sad" state where we might want to allow an operator/VM user to examine the state of the VM more closely (see, for example, #340). I would be inclined to treat these as separate conditions though ("Failed" vs "GuestFailed" or something like that).

I'm a little concerned that destroying the VM on a startup failure (without requiring a request to stop) is too aggressive of an approach. If others are similarly concerned I can try weakening that a bit.

gjcolombo commented:
I believe Buildomat is taking a nap, so the PHD run for these changes is on hold. I'll wait until it's back before merging anything here.

luqmana commented on bin/propolis-server/src/lib/vm/state_driver.rs:
```diff
@@ -798,13 +830,99 @@ mod tests {
+        // queue a "stop" command to itself in this case, but because the driver
+        // is not directly processing the queue here, the test has to issue this
+        // call itself.
+        assert!(matches!(
+            driver.driver.get_migration_state(),
+            Some((_, ApiMigrationState::Finish))
+        ));
+        assert_eq!(
+            driver.driver.get_migration_state().unwrap().0,
+            migration_id
+        );
```
You can fold these together with a guard on the `matches!`:

```suggestion
        assert!(matches!(
            driver.driver.get_migration_state(),
            Some((id, ApiMigrationState::Finish)) if id == migration_id
        ));
```

gjcolombo replied:
Good call. Done in 3a2c071.

luqmana reviewed (commented):
>I'm a little concerned that destroying the VM on a startup failure (without requiring a request to stop) is too aggressive of an approach. If others are similarly concerned I can try weakening that a bit.

I'm inclined to agree, I think we should we should mark it as `Failed` but leave the vmm state available. Sled Agent or whatever other consumer can choose to further clean it up.

gjcolombo commented:
> I'm inclined to agree, I think we should we should mark it as Failed but leave the vmm state available. Sled Agent or whatever other consumer can choose to further clean it up.

I gave this a bit more thought. It still seems a bit odd to me to preserve the `Instance` in cases where we didn't actually try to start it (e.g. because of a failure in the migration protocol), but I've come around to thinking it's valuable to have in cases where we fail to start the VM and want to have the option of writing a dump and debugging it. So I changed this back in c724f0b.

We could also split the difference here and return `HandleEventOutcome::Exit` from migration protocol failures and `HandleEventOutcome::Continue` from VM start failures. I didn't do that for now just to keep things consistent. If it continues to be a big pain point for migration testing, we can always revisit (my WIP sled agent changes don't currently care about the behavior here so we have time to settle on what it should be). CC @jordanhendricks on this point since it came up in the discussion on #323.

gjcolombo commented:
Thanks for the review!
